### DISCUSSION TITLE
Support for conversion between bytes and std::string

### DISCUSSION THREAD

**Discussion by brendan-ward** [wjakob/nanobind#137] [Q&A] [answered]
pybind11 supported automatic conversion between [`bytes` and `std::string`](https://pybind11.readthedocs.io/en/stable/advanced/cast/strings.html#passing-bytes-to-c)

nanobind supports conversion from `bytes` to `char *` (added in #62) but not (yet?) `std::string`.

Some support for this was drafted in #23, but that was closed without merging, and the implementation and goals (if I understood them correctly) were a bit different than the approach used in #62.

I'm unable to tell if the lack of automatic conversion in nanobind is an intentional design choice (users responsible for managing their own conversions) or if it simply hasn't been implemented yet (i.e., PR would be welcome).  My interest would be that this works just like in pybind11 (no conversion of underlying string data / encoding), since this is a point of friction in migrating from pybind11 to nanobind.

**Comment by wjakob** [comment, marked as answer]
A correction to what you said: commit #62 introduced a *wrapper type* (see this page for details: https://nanobind.readthedocs.io/en/latest/exchanging.html). What you ask for is a *type converter* that automatically casts `std::string` (or potentially, `char *` -- same thing) into `bytes` and vice versa. pybind11 really only supported this because of its Python 2.x heritage involving a mixture of byte and unicode-format strings. I don't actually think it's a good idea to do this anymore, and I am not eager to add support for it. My suggestion would be that you stay with pybind11, create a custom fork of nanobind, or adapt your own binding interface.

**Reply by brendan-ward** [reply]
> adapt your own binding interface

I ended up doing this.  The switch to nanobind has otherwise been fairly smooth, and once I figured out the correct (I hope) way of converting, it was only a few lines of code.  It's helpful to know that it isn't desirable to have this particular automatic conversion as part of nanobind.
